controller/src/main/java/io/pravega/controller/server/SegmentHelper.java

```diff
@@ -1303,6 +1086,52 @@ private boolean hasConnectionFailed(Throwable unwrap) {
         return new ImmutablePair<>(rateType, desiredRate);
     }
 
+    private void closeConnection(Reply reply, RawClient client) {
+        log.info("Closing connection as a result of receiving: {}", reply);
+        if (client != null) {
+            try {
+                client.close();
+            } catch (Exception e) {
+                log.warn("Exception tearing down connection: ", e);
+            }
+        }
+    }
+
+    /**
+     * This method handle reply returned from RawClient.sendRequest.
+     *
+     * @param reply               actual reply received
+     * @param expectedReplies     set of the expected replies
+     * @param client              RawClient for sending request
+     * @return true if reply is in the expected reply set or throw exception.
+     */
+    private Boolean transformReply(Reply reply,
+                                   Map<? extends Class, Runnable> expectedReplies,
+                                   RawClient client) {
+        if (expectedReplies.containsKey(reply.getClass())) {
+            expectedReplies.get(reply.getClass()).run();
+            closeConnection(reply, client);
+            return true;
+        } else {
+            throw handleUnexpectedReply(reply, expectedReplies, client);
+        }
+    }
+
+    @SneakyThrows(ConnectionFailedException.class)
+    private RuntimeException handleUnexpectedReply(Reply reply,
+                                                   Map<? extends Class, Runnable> expectedReplies,
+                                                   RawClient client) {
+        closeConnection(reply, client);
+        if (reply instanceof WireCommands.NoSuchSegment) {
+            throw new NoSuchSegmentException(reply.toString());
+        } else if (reply instanceof WireCommands.WrongHost) {
+            throw new ConnectionFailedException(reply.toString());
+        } else {
+            throw new ConnectionFailedException("Unexpected reply of " + reply + " when expecting one of "
+                    + expectedReplies.keySet().toString());
```

Comment:
Are you sure this will have the output you expect? I suggest using `String.join` to compose a proper String.

Reply:
tested, it works, an extra test case also added in SegmentHelperTest.java

```at java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:273)
	at java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:280)
	at java.util.concurrent.CompletableFuture.uniApply(CompletableFuture.java:604)
	at java.util.concurrent.CompletableFuture$UniApply.tryFire(CompletableFuture.java:577)
	at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:474)
	at java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:1962)
	at io.pravega.client.netty.impl.RawClient.reply(RawClient.java:104)
	...
	at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:46)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:55)
	at java.lang.Thread.run(Thread.java:748)
Caused by: io.pravega.shared.protocol.netty.ConnectionFailedException: Unexpected reply of WireCommands.SegmentDeleted(type=SEGMENT_DELETED, requestId=17179869184, segment=/0.#epoch.0) when expecting one of [class io.pravega.shared.protocol.netty.WireCommands$SegmentAlreadyExists, class io.pravega.shared.protocol.netty.WireCommands$SegmentCreated]```